lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1417 @@
+    } -ErrorMessage "Failed to run pywin32_postinstall.py"
+}
+
+
+function Install-FreeRDPConsole {
+    Write-JujuLog "Installing FreeRDP"
+
+    Install-VC2012
+
+    $freeRDPZip = Join-Path $FILES_DIR "FreeRDP_powershell.zip"
+    $charmLibDir = Join-Path (Get-JujuCharmDir) "lib"
+    Expand-ZipArchive $freeRDPZip $charmLibDir
+
+    # Copy wfreerdp.exe and DLL file to Windows folder
+    $freeRDPFiles = @('wfreerdp.exe', 'libeay32.dll', 'ssleay32.dll')
+    $windows = Join-Path $env:SystemDrive "Windows"
+    foreach ($file in $freeRDPFiles) {
+        Copy-Item "$charmLibDir\FreeRDP\$file" $windows
+    }
+
+    $freeRDPModuleFolder = Join-Path $windows "system32\WindowsPowerShell\v1.0\Modules\FreeRDP"
+    if (!(Test-Path $freeRDPModuleFolder)) {
+        mkdir $freeRDPModuleFolder
+    }
+    Copy-Item "$charmLibDir\FreeRDP\FreeRDP.psm1" $freeRDPModuleFolder
+    Remove-Item -Recurse "$charmLibDir\FreeRDP"
+
+    Write-JujuLog "Finished installing FreeRDP."
+}
+
+
+function Write-PipConfigFile {
+    $pipDir = Join-Path $env:APPDATA "pip"
+    if (Test-Path $pipDir){
+        Remove-Item -Force -Recurse $pipDir
+    }
+
+    $pypiMirror = Get-JujuCharmConfig -scope 'pypi-mirror'
+    if ($pypiMirror -eq $null -or $pypiMirror.Length -eq 0) {
+        Write-JujuLog ("pypi-mirror config is not present. " +
+                       "Will not generate the pip.ini file.")
+        return
+    }
+    mkdir $pipDir
+    $pipIni = Join-Path $pipDir "pip.ini"
+    New-Item -ItemType File $pipIni
+
+    $mirrors = $pypiMirror.Split()
+    $hosts = @()
+    foreach ($i in $mirrors){
+        $h = ([System.Uri]$i).Host
+        if ($h -in $hosts) {
+            continue
+        }
+        $hosts += $h
+    }
+
+    Set-IniFileValue "index-url" "global" $mirrors[0] $pipIni
+    if ($mirrors.Length -gt 1){
+        Set-IniFileValue "extra-index-url" "global" ($mirrors[1..$mirrors.Length] -Join " ") $pipIni
+    }
+    Set-IniFileValue "trusted-host" "install" ($hosts -Join " ") $pipIni
+}
+
+
+function Get-HypervADUser {
+    $adUsername = Get-JujuCharmConfig -scope 'ad-user-name'
+    if (!$adUsername) {
+        $adUsername = "hyper-v-user"
+    }
+    return $adUsername
+}
+
+
+function Set-DevStackRelationParams {
+    Param(
+        [HashTable]$RelationParams
+    )
+
+    $rids = Get-JujuRelationIds -Relation "devstack"
+    foreach ($rid in $rids) {
+        try {
+            Set-JujuRelation -Settings $RelationParams -RelationId $rid
+        } catch {
+            Write-JujuError "Failed to set DevStack relation parameters."
+        }
+    }
+}
+
+
+function Import-CloudbaseCert {
+    $crt = Join-Path $FILES_DIR "Cloudbase_signing.cer"
+    if (!(Test-Path $crt)){
+        return $false
+    }
+    Import-Certificate $crt -StoreLocation LocalMachine -StoreName TrustedPublisher
+}
+
+
+function Set-AD2DevstackCreds {
+    $adUserCred = @{
+        'domain'   = $adCtx["domainName"];
```

Comment:
I don't see any `$adCtx` variable in this function. Unless you initialized `$adCtx` variable before calling this function, there is no way that this will work.

Please get the context for AD at the beginning of the function and check if it was the required variables, otherwise the `Set-JujuRelation` will fail.